### PR TITLE
Initial pull request for first version

### PR DESCRIPTION
**Pull request by davidcolton**
Bob. Hopefully I got this right.

### CODE REVIEW

**Comment by davidcolton** [comment]
@bbelderbos I can't seem to add specific reviewers.

File: content/guest-cleaning-text-data.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+8. [Stemming and Lemmatisation](#stemming-lemmatisation)
+9. [A Simple Demonstration](#demonstration)
+
+<a name="cleaning-text"></a>
+
+# Cleaning Text Data with Python
+
+Machine Learning is super powerful if your data is numeric. What do you do, however, if you want to mine text data to discover hidden insights or to predict the sentiment of the text. What, for example, if you wanted to identify a post on a social media site as cyber bullying. 
+
+One way to analyse text is to use a measure called Term Frequency - Inverse Document Frequency (TF-IDF). A full explanation of TF-IDF is beyond the scope of this quick introduction, however, a detailed background and explanation of TF-IDF, including some Python examples, is given here [Analyzing Documents with TF-IDF](https://programminghistorian.org/en/lessons/analyzing-documents-with-tfidf). Suffice to say that TF-IDF will assign a value to every word in every document you want to analyse and, the higher the TF-IDF value, the more important or predictive the word will typically be.
```
Comment (jazcap53) [inline content/guest-cleaning-text-data.md:28]: "Suffice to say" should be "Suffice it to say"

**Comment by bbelderbos** [comment]
> @bbelderbos I can't seem to add specific reviewers.

Yeah it seems for that you need to be collaborator on the repo. Nws, I like the article, interesting. I will push it out and we can always make edits as Bites come out etc. Thanks!

File: content/guest-cleaning-text-data.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+* _this movie was not good_
+* _movie good_
+
+By removing stop words you've changed the sentiment of the sentence. Who said NLP and Text Mining was easy.
+
+<a name="word-standardisation"></a>
+
+## Spelling and Repeated Characters (Word Standardisation)
+
+Fixing obvious spelling errors can both increase the predictiveness of you model and speed up processing by reducing the size of your corpora. A good example of this is on Social Media sites when words are either truncated, deliberately misspelt or accentuated by adding unnecessary repeated characters. Consider:
```
Comment (pmayd) [inline content/guest-cleaning-text-data.md:67]: "of you model" should be "of your model".

File: content/guest-cleaning-text-data.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+* `love, luv, lovvvvv, lovvveeee`
+
+To an English speaker it pretty obvious that the single word the represents all these tokens is `love`. Standardising your text in this manner has the potential to improve your model significantly.
```
Comment (pmayd) [inline content/guest-cleaning-text-data.md:71]: "it pretty obvious" should be "it's pretty obvious". 
"the represents" should be " that represents".

You say that this is important but how do you achieve such a standardization?

File: content/guest-cleaning-text-data.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+* `love, luv, lovvvvv, lovvveeee`
+
+To an English speaker it pretty obvious that the single word the represents all these tokens is `love`. Standardising your text in this manner has the potential to improve your model significantly. 
+
+<a name="remove"></a>
+
+## Remove URL's, Email Addresses and Emojis
+
+Depending on your modelling requirements your might want to either leave these items in you text or further preprocess them as required. A general approach though is to assume these are not required and should be excluded. Consider if is it worth converting your emojis to text, would this bring extra predictiveness to your model. Regular expressions are the go to solution for removing URL's and email addresses.
```
Comment (pmayd) [inline content/guest-cleaning-text-data.md:77]: "your might want" should be "you might want".
"you text" should be "your text".
"Consider if is it" should be "Consider if it is ".
The sentence "Consider if is it.." is a question so add a ? at the end?
"remmoving URL's" should be "removing URLs", its plural not possession

File: content/guest-cleaning-text-data.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+<a name="stemming-lemmatisation"></a>
+
+## Stemming and Lemmatisation
+
+Stemming is a process by which derived or inflected words are reduced to their stem, sometimes also called the base or root. Using the words `stemming` and `stemmed` as examples, these are both based on the word `stem`. **Stemming** algorithms work by cutting off the end or the beginning of the word, taking into account a list of common prefixes and suffixes that can be found in an inflected word.
+
+**Lemmatisation** in linguistics, is the process of grouping together the different inflected forms of a word so they can be analysed as a single item. In languages, words can appear in several inflected forms. For example, in English, the verb 'to walk' may appear as 'walk', 'walked', 'walks', 'walking'. The base form, 'walk', that one might look up in a dictionary, is called the lemma for the word.
+
+So stemming uses *predefined rules* to transform the word into a *stem* whereas lemmatisation uses *context* and *lexical library* to derive *lemma*. *Stem* doesn’t always have to be a valid word whereas *lemma* will always be a valid word because *lemma* is a dictionary form of a word.
```
Comment (pmayd) [inline content/guest-cleaning-text-data.md:87]: "to derive lemma" missing an article like "to derive a lemma" or the.
Same for "sStem doesn't always" -> "A/The Stem doesn't always" and lemma

**Comment by bbelderbos** [comment]
Thanks @pmayd and @jazcap53 for reviewing. 
I had already merged it so I made the spelling fixes on the main branch.
@davidcolton as there are still some unresolved conversations I put the article in draft status for now.

File: content/guest-cleaning-text-data.md (hunk 1 above)
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:28]: Updated

[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:28]: Done.

File: content/guest-cleaning-text-data.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+One way to analyse text is to use a measure called Term Frequency - Inverse Document Frequency (TF-IDF). A full explanation of TF-IDF is beyond the scope of this quick introduction, however, a detailed background and explanation of TF-IDF, including some Python examples, is given here [Analyzing Documents with TF-IDF](https://programminghistorian.org/en/lessons/analyzing-documents-with-tfidf). Suffice to say that TF-IDF will assign a value to every word in every document you want to analyse and, the higher the TF-IDF value, the more important or predictive the word will typically be.
+
+However, before you can use TD-IDF you need to clean up your text data. But why do we need to clean text, can we not just eat it straight out of the tin. The answer is yes you can, you can use the raw data exactly as you've received it, however, cleaning your data will increase the accuracy of your model. This guide is a very basic introduction to some of the approaches used in cleaning text data. Some techniques are simple, some more advanced. For the more advanced concepts, consider their inclusion here as pointers for further personal research.
```
[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:30]: Appreciate what you are saying here @pmayd about not explaining why cleaning the text will give better results but at the end of the day this is just a small introduction to methods that can be used to clean text. Maybe a good topic for the ML learning path if you get it going.

Minor issues cleaned up though.

File: content/guest-cleaning-text-data.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+However, before you can use TD-IDF you need to clean up your text data. But why do we need to clean text, can we not just eat it straight out of the tin. The answer is yes you can, you can use the raw data exactly as you've received it, however, cleaning your data will increase the accuracy of your model. This guide is a very basic introduction to some of the approaches used in cleaning text data. Some techniques are simple, some more advanced. For the more advanced concepts, consider their inclusion here as pointers for further personal research. 
+
+In the following sections I'm assuming that you have plain text and your text is not embedded in HTML or Markdown or anything like that. If is is you should handle this first to get access to the raw text before proceeding.
```
[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:32]: Done

File: content/guest-cleaning-text-data.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+<a name="normalising-case"></a>
+
+## Normalising Case
+
+This is just a fancy way of saying convert all your text to lowercase. If using Tf-IDF `Hello` and `hello` are two different tokens. This has the side affect or reducing the total size of the vocabulary, or corpus, and some knowledge will be lost such as Apple the company versus eating an apple. In all cases you should consider if each of these actions actually makes sense to the text analysis you are performing. If you are not sure, or you want to see the impact of a particular cleaning technique try the before and after text to see which approach gives you a more predictive model. Sometimes, in text mining, there are multiple right answers.
```
[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:44]: Done

File: content/guest-cleaning-text-data.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+<a name="remove-punctuation"></a>
+
+## Remove All Punctuation
+
+Punctuation doesn't bring anything to the table when text mining so just remove it all. In fact sentence structure and word order is irrelevant when using TD-IDF.  Word of caution though. If you are also going to remove URL's and Email addresses you might want to the do that before removing punctuation characters otherwise they'll be a bit hard to identify. Another consideration is hashtags which you might want to keep so you may need a rule to remove `#` unless it is the first character of the token.
```
[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:50]: Agreed, too broad, but I do allude to this in stop word removal.

[the remark this answers is not in the document]
Comment (davidcolton) [reply content/guest-cleaning-text-data.md:50]: I was hoping to avoid having to speak about bag of words vs NLP. Maybe I can't avoid it now.

**Pull request merged**